### PR TITLE
[Feature] Add points_in_polygons CUDA op for rotated detection. 

### PR DESCRIPTION
**Pull request by zytx121**
Add points_in_polygons CUDA ops for anchor-free rotated detection. It is mainly used to judge whether the points are inside the ground true polygons during sample assignment.

## Modification

Several files in mmcv/ops folder.

## BC-breaking (Optional)

No.

## Use cases (Optional)

from mmcv.ops import points_in_polygons

### CODE REVIEW

**Comment by zhouzaida** [comment]
Please @AllentDan have a look.

**Comment by AllentDan** [comment]
Hi, @zytx121. Unit tests failed on my device.

**Comment by zytx121** [comment]
> Hi, @zytx121. Unit tests failed on my device.

Sorry, I'll check it as soon as possible.

**Comment by zytx121** [comment]
> Hi, @zytx121. Unit tests failed on my device.

Please test again. Thx.

**Review by AllentDan — APPROVED** [review]
LGTM except updating ops in https://github.com/open-mmlab/mmcv/blob/master/docs/en/understand_mmcv/ops.md and https://github.com/open-mmlab/mmcv/blob/master/docs/zh_cn/understand_mmcv/ops.md

File: mmcv/ops/points_in_polygons.py [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+    Args:
+        points (torch.Tensor): It has shape (B, 2), indicating (x, y).
+        polygons (torch.Tensor): It has shape (M, 8), indicating
+            (x1, y1, x2, y2, x3, y3, x4, y4).
+
+    Returns:
+        torch.Tensor: Return the result with the shape of (B, M),
+            1 indicates that the point is inside the polygon,
+            0 indicates that the point is outside the polygon.
```
Comment (zhouzaida) [inline mmcv/ops/points_in_polygons.py:19]:
```suggestion
        torch.Tensor: Return the result with the shape of (B, M),
        1 indicates that the point is inside the polygon,
        0 indicates that the point is outside the polygon.
```

File: tests/test_ops/test_points_in_polygons.py [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+@pytest.mark.skipif(
+    not torch.cuda.is_available(), reason='requires CUDA support')
+def test_points_in_polygons():
+    points = np.array([[300., 300.], [400., 400.], [100., 100], [300, 250],
+                       [100, 0]])
+    polygons = np.array([[200., 200., 400., 400., 500., 200., 400., 100.],
+                         [400., 400., 500., 500., 600., 300., 500., 200.],
+                         [300., 300., 600., 700., 700., 700., 700., 100.]])
+    expert_output = np.array([[0., 0., 0.], [0., 0., 1.], [0., 0., 0.],
```
Comment (zhouzaida) [inline tests/test_ops/test_points_in_polygons.py:16]:
```suggestion
    expected_output = np.array([[0., 0., 0.], [0., 0., 1.], [0., 0., 0.],
```

File: mmcv/ops/points_in_polygons.py (hunk 1 above)
Comment (zhouzaida) [reply mmcv/ops/points_in_polygons.py:19]: Perhaps we can explain more specifically the meaning of the dimension M.

Comment (zytx121) [reply mmcv/ops/points_in_polygons.py]: OK

**Comment by grimoire** [comment]
Is there any limit about the polygon? Such as convex or clockwise?

Test failed with data:

```python
    points = np.array([[20, 90]])
    polygons = np.array([[0., 0., 100., 100., 10., 90., 0., 100.]])
```

If there are any limit about the polygon, please update the docstring.

**Comment by zytx121** [comment]
> Is there any limit about the polygon? Such as convex or clockwise?
> 
> Test failed with data:
> 
> ```python
>     points = np.array([[20, 90]])
>     polygons = np.array([[0., 0., 100., 100., 10., 90., 0., 100.]])
> ```
> 
> If there are any limit about the polygon, please update the docstring.

We believe that this method can handle all polygons without adding conditions such as convex or clockwise. As shown in the figure below, this point falls inside the polygon, which is consistent with the test results. 

image

```pyhon
def test_points_in_polygons():
    points = np.array([[20, 90]])
    polygons = np.array([[0., 0., 100., 100., 10., 90., 0., 100.]])
    expert_output = np.array([[1.,]])
    points = torch.from_numpy(points).cuda().float()
    polygons = torch.from_numpy(polygons).cuda().float()
    expert_output = torch.from_numpy(expert_output).cuda().float()
    assert torch.allclose(
        points_in_polygons(points, polygons), expert_output, 1e-3)

```

Note that the format of the polygon is (x,y,x,y,x,y,x,y).

**Comment by grimoire** [comment]
@zytx121 Thanks for the explanation. The data did pass the test as you said.

**Review by grimoire — APPROVED** [review]
LGTM.

File: tests/test_ops/test_points_in_polygons.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+    points = np.array([[300., 300.], [400., 400.], [100., 100], [300, 250],
+                       [100, 0]])
+    polygons = np.array([[200., 200., 400., 400., 500., 200., 400., 100.],
+                         [400., 400., 500., 500., 600., 300., 500., 200.],
+                         [300., 300., 600., 700., 700., 700., 700., 100.]])
+    expected_output = np.array([[0., 0., 0.], [0., 0., 1.], [0., 0., 0.],
+                                [1., 0., 0.], [0., 0., 0.]])
+    points = torch.from_numpy(points).cuda().float()
+    polygons = torch.from_numpy(polygons).cuda().float()
+    expert_output = torch.from_numpy(expert_output).cuda().float()
```
Comment (zhouzaida) [inline tests/test_ops/test_points_in_polygons.py:20]:
```suggestion
    expert_output = torch.from_numpy(expected_output).cuda().float()
```

File: tests/test_ops/test_points_in_polygons.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    assert torch.allclose(
+        points_in_polygons(points, polygons), expert_output, 1e-3)
```
Comment (zhouzaida) [inline tests/test_ops/test_points_in_polygons.py:22]:
```suggestion
        points_in_polygons(points, polygons), expected_output, 1e-3)
```

**Review by zhouzaida — APPROVED** [review]

File: mmcv/ops/csrc/pytorch/cuda/points_in_polygons_cuda.cu [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+
+void PointsInPolygonsForwardCUDAKernelLauncher(const at::Tensor points,
+                                               const at::Tensor polygons,
+                                               const int rows, const int cols,
+                                               at::Tensor output) {
+  const int output_size = rows * cols;
+  at::cuda::CUDAGuard device_guard(points.device());
+  cudaStream_t stream = at::cuda::getCurrentCUDAStream();
+  AT_DISPATCH_FLOATING_TYPES_AND_HALF(
+      points.scalar_type(), "PointsInPolygonsLaucher", ([&] {
```
Comment (zhouzaida) [inline mmcv/ops/csrc/pytorch/cuda/points_in_polygons_cuda.cu:18]:
```suggestion
      points.scalar_type(), "points_in_polygons_forward_cuda_kernel", ([&] {
```

**Review by zhouzaida — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]

**Comment by ZwwWayne** [comment]
Can be merged after resolving conflicts.

**Pull request merged**